Feat: Add KPI/Rewarder Module

0xNuggan opened the pull request.

FHieser committed:
344 staking contract v2 (#382)

* Create IStakingManager.sol

* Create StakingManager and adapt Interface

* Finalizing stakingManager

* Make it buildable

* Rename Stakingmanager

* Create StakingManager.t.sol

* Create StakingManagerAccessMock.sol

* further testing

* Test distribute Rewards

* test withdraw

* test setRewards

* Complete tests

* Implement e2e test

* pre-commit

* Create IStakingManager.sol

* Create StakingManager and adapt Interface

* Finalizing stakingManager

* Make it buildable

* Rename Stakingmanager

* Create StakingManager.t.sol

* Create StakingManagerAccessMock.sol

* further testing

* Test distribute Rewards

* test withdraw

* test setRewards

* Complete tests

* Implement e2e test

* pre-commit

* Remove todo

* pre-commit

* Update Documentation

0xNuggan committed:
Uma optimistic oracle mixin (#383)

* forge install: protocol

54fbb49

* create UMA Oracle Integrator base contract

* partially resolve dependency issue

* Test skeleton

* first versions of mocks

* wip tests

* fix setup

* WIP tests

* first batch of tests

* forge install: protocol

54fbb49

* library changes

* forge install: protocol

7b91963

* refactor imports

* format

* revert full migration to UMA library

Since we run into solidity version issues, I undid the migration there. We copy some core UMA contracts into our folder to change the solidity version

* clean up

* format

* Delete EscalationManagerInterface.sol

* fix TODOs

0xNuggan committed:
KPI Rewarder draft

0xNuggan committed:
comment

0xNuggan committed:
Update KPIRewarder.sol

0xNuggan committed:
rename withdraw to unstake

0xNuggan committed:
Update KPIRewarder.sol

0xNuggan committed:
outlin in comments

0xNuggan committed:
small StakingManager refactor

0xNuggan committed:
format

0xNuggan committed:
more dev + set up test

0xNuggan committed:
generate Tree scaffolds

0xNuggan committed:
continue test writing

0xNuggan committed:
extract interface

0xNuggan committed:
expand event  info + refactoring

0xNuggan committed:
review bond payment

0xNuggan committed:
more test dev

0xNuggan committed:
more tests

0xNuggan committed:
more tests

0xNuggan committed:
finish tests, missing cleanup

0xNuggan committed:
hotifx

FHieser committed:
Hope it helps

0xNuggan committed:
update from felix

0xNuggan committed:
natspec

0xNuggan committed:
refactor user generation + test reward emission event

0xNuggan committed:
ERC165 comments

FHieser committed:
Fuzz Rewards

FHieser committed:
Corrections

FHieser commented on test/modules/logicModule/KPIRewarder.t.sol:
```diff
@@ -0,0 +1,1040 @@
+            assertEq(kpiManager.balanceOf(users[i]), amounts[i]);
+
+            //=========================================================
+            // This is the place where imprecision issues arise. Needs review
```
@0xNuggan Are you really sure this problem lies in the stakingmanager?
In the last commit I added fuzzing to the reward amounts and it went through the tests without problems
Also the only place this seems to be the problem is the continous section. Everywhere else it doesnt throw any errors right?
I would suggest looking at that part again. Especially the interaction between the estimateReward function from the StakingManager and the the Tranches in general. Especially because you didnt override it in the KPI rewarder

0xNuggan committed:
additional comments

FHieser committed:
Fix estimate reward

FHieser committed:
fix comments

FHieser committed:
Update KPIRewarder.t.sol

FHieser committed:
Revert to comment

FHieser commented on src/modules/logicModule/IKPIRewarder.sol:
```diff
@@ -0,0 +1,151 @@
+    /// @notice Creates a KPI for the Rewarder
+    /// @param _continuous Should the tranche rewards be distributed continuously or in steps
+    /// @param _trancheValues The value at which the tranches end
+    /// @param _trancheRewards The rewards to be diitributed at completion of each tranche
```
*distributed

FHieser committed:
Fix testEstimateReward

FHieser committed:
Add comments

FHieser commented on src/modules/logicModule/KPIRewarder.sol:
```diff
@@ -0,0 +1,365 @@
+        setKPI(targetKPI);
+        _setAssertion(dataId, data, asserter);
+    }
+
```
/// @inheritdoc IKPIRewarder

FHieser commented on src/modules/logicModule/KPIRewarder.sol:
```diff
@@ -0,0 +1,365 @@
+    }
+
+    /// @inheritdoc IKPIRewarder
+    function setKPI(uint _KPINumber) public onlyOrchestratorOwner {
```
What is the point of having this function if you have prepareAssertion?

FHieser commented on src/modules/logicModule/KPIRewarder.sol:
```diff
@@ -0,0 +1,365 @@
+    }
+
+    /// @inheritdoc IKPIRewarder
+    function setActiveTargetValue(uint targetValue)
```
What is the point of having this function if you have prepareAssertion?

FHieser commented on src/modules/logicModule/KPIRewarder.sol:
```diff
@@ -0,0 +1,365 @@
+            // SECURITY NOTE: this will add the value, but provides no guarantee that the fundingmanager actually holds those funds.
+
+            // Calculate rewardamount from assertionId value
+            KPI memory resolvedKPI =
```
Why do we have a Active KPI if you dont trigger it here?

FHieser commented on src/modules/logicModule/KPIRewarder.sol:
```diff
@@ -0,0 +1,365 @@
+    /// @dev about the asserter address: any address can be set as asserter, it will be expected to pay for the bond on posting.
+    /// The bond tokens can also be deposited in the Module and used to pay for itself, but ONLY if the bond token is different from the one being used for staking.
+    /// If the asserter is set to 0, whomever calls postAssertion will be paying the bond.
+    function prepareAssertion(
```
Why do we need an active assertion in general here?
Technically we could have multiple different ones at the same time.
We could just create a AssertionStruct that saves all the accoring data and then sent it in the postAssertion field

FHieser reviewed (commented):
From an implementation logic standpoint it looks good I think,
but there is definitely a how to guide needed. There are a lot of keywords flying around here and bringing them in order via a guide would be pretty great.

Also I think we could actually remove the "active" Assertion part (including active KPI, targetValue etc) in here and just keep them saved as a Assertion struct for later reference.   Technically these assertions could be (and I think actually currently can be) run in Parallel if you wanted them. The declaration of them being "Active" doesn't really mean much if I understood it correctly.

0xNuggan committed:
hotfixes

0xNuggan committed:
simplify Rewarder Structure

removed the concept of "active" KPI or asserions. They now have to be supplied in the postAssertion() call

0xNuggan committed:
adapt tests

0xNuggan committed:
Add explanation text

0xNuggan committed:
add dequeue() function

FHieser reviewed (commented):
Nice cleanup.
I think that makes the code pretty tight.
Do we still have the issues on the tests?

0xNuggan committed:
typo

0xNuggan committed:
updates on precision issue

FHieser committed:
Update KPIRewarder.t.sol

0xNuggan committed:
Minmum test pass
